### PR TITLE
Add optimizer scheduler and improve diffusion policy

### PR DESCRIPTION
**Pull request by kwangneuraco**
### Features

- Add optimizer scheduler across all algorithms
-  Improved Diffusion Policy especially hyper-parameters, pre-trained network loading

Related Tickets:
[NCRL-72](https://neuracore.atlassian.net/browse/NCRL-72?atlOrigin=eyJpIjoiMGEyZDk1OWVhMWFlNDZlOWJlN2FhZDdlOWU4OTk2YWYiLCJwIjoiaiJ9), [NCRL-31](https://neuracore.atlassian.net/browse/NCRL-31)

### CODE REVIEW

File: neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py [hunk 1]
```diff
@@ -581,12 +616,52 @@ def configure_optimizers(self) -> list[torch.optim.Optimizer]:
...
-            {"params": backbone_params, "lr": self.lr_backbone},
-            {"params": other_params, "lr": self.lr},
-        ]
+        # Build parameter groups, filtering out empty ones
+        param_groups = []
+
+        # If lr_backbone is 0, freeze backbone parameters and exclude from optimizer
+        if self.lr_backbone == 0:
+            for param in backbone_params:
+                param.requires_grad = False
```
Comment (ypang-neuraco) [inline neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py:626]: Should this logic be added to other models as well?

File: neuracore/ml/algorithms/cnnmlp/cnnmlp.py [hunk 2]
```diff
@@ -517,14 +517,23 @@ def training_step(self, batch: BatchedTrainingSamples) -> BatchedTrainingOutputs
             metrics=metrics,
         )
 
-    def configure_optimizers(self) -> list[torch.optim.Optimizer]:
+    def configure_optimizers(
+        self,
+        num_training_steps: Optional[int] = None,
```
Comment (ypang-neuraco) [inline neuracore/ml/algorithms/cnnmlp/cnnmlp.py:522]: I think this is argument is specific to the scheduler used in diffusion policy so it should not be added to other algorithms. Perhaps use a generic "optimizer_config" dictionary instead.

File: neuracore/ml/trainers/batch_autotuner.py [hunk 3]
```diff
@@ -51,7 +51,17 @@ def __init__(
         self.model = model
 
         # create optimizers
-        self.optimizers = self.model.configure_optimizers()
+        optimizer_result = self.model.configure_optimizers()
+        if isinstance(optimizer_result, dict):
```
Comment (ypang-neuraco) [inline neuracore/ml/trainers/batch_autotuner.py:55]: This check seems redundant, maybe you can create a unit test to test for this in each algorithm

File: neuracore/ml/trainers/distributed_trainer.py [hunk 4]
```diff
@@ -115,10 +117,24 @@ def __init__(
...
+            optimizers = optimizer_result.get("optimizers")
+            if optimizers is None:
+                raise ValueError("optimizers cannot be None")
+            self.optimizers: list[Optimizer] = optimizers
+            self.schedulers: Optional[list] = optimizer_result.get("schedulers")
+        else:
+            raise ValueError(
+                "configure_optimizers must return a dictionary with keys "
+                "'optimizers' and 'schedulers'"
+            )
```
Comment (ypang-neuraco) [inline neuracore/ml/trainers/distributed_trainer.py:137]: I don't think we need these checks, maybe only keep the one for optimizer not None

File: examples/ml/example_custom_algorithm.py [hunk 5]
```diff
@@ -620,7 +628,12 @@ def configure_optimizers(self) -> list[torch.optim.Optimizer]:
             {"params": backbone_params, "lr": self.lr_backbone},
             {"params": other_params, "lr": self.lr},
         ]
-        return [torch.optim.AdamW(param_groups, weight_decay=self.weight_decay)]
+        return {
+            "optimizers": [
+                torch.optim.AdamW(param_groups, weight_decay=self.weight_decay)
+            ],
+            "schedulers": None,
```
Comment (ypang-neuraco) [inline examples/ml/example_custom_algorithm.py:635]: empty list instead of None? Can also add type LRScheduler in the output

Comment (stepjam) [reply examples/ml/example_custom_algorithm.py:635]: agreed

File: examples/ml/example_custom_algorithm.py [hunk 6]
```diff
@@ -595,14 +595,22 @@ def training_step(self, batch: BatchedTrainingSamples) -> BatchedTrainingOutputs
             metrics=metrics,
         )
 
-    def configure_optimizers(self) -> list[torch.optim.Optimizer]:
+    def configure_optimizers(
```
Comment (stepjam) [inline examples/ml/example_custom_algorithm.py:598]: Considering that optimizers do not need num_training_steps, I suggest you split this function, and we have a `configure_optimizers`, and a `configure_schedulers(optimizers: list[torch.optim.Optimizer]`

File: neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py [hunk 7]
```diff
@@ -581,12 +617,50 @@ def configure_optimizers(self) -> list[torch.optim.Optimizer]:
             else:
                 other_params.append(param)
 
-        param_groups = [
-            {"params": backbone_params, "lr": self.lr_backbone},
-            {"params": other_params, "lr": self.lr},
-        ]
+        # Build parameter groups, filtering out empty ones
```
Comment (stepjam) [inline neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py:620]: This will look less bloated once the oprimizers and schedulers are split (previous comment)

Comment (stepjam) [reply neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py:620]: Also, a lot of the code below is just not necessary. Surely if lr_backbone is 0, then we can just not construct a backbone optimizer? I.e.:
```
param_groups = [
       {"params": other_params, "lr": self.lr},
]
```

rather than:
```
param_groups = [
      {"params": backbone_params, "lr": self.lr_backbone},
       {"params": other_params, "lr": self.lr},
]
```

File: neuracore/ml/trainers/distributed_trainer.py [hunk 8]
```diff
@@ -115,10 +117,24 @@ def __init__(
...
         self.rank = rank
         self.world_size = world_size
-        self.optimizers = model.configure_optimizers()
         self.global_train_step = 0
         self.global_val_step = 0
 
+        num_training_steps = self.num_epochs * len(self.train_loader)
+        optimizer_result = model.configure_optimizers(
+            num_training_steps=num_training_steps
+        )
```
Comment (stepjam) [inline neuracore/ml/trainers/distributed_trainer.py:126]: As notes above, configure_optimizers should not have num_training_steps

File: neuracore/ml/utils/training_storage_handler.py [hunk 9]
```diff
@@ -134,6 +138,34 @@ def save_checkpoint(self, checkpoint: dict, relative_checkpoint_path: Path) -> N
                 )
                 return
 
+    def _convert_omegaconf_to_python(self, obj: Any) -> Any:
+        """Recursively convert OmegaConf objects to plain Python types.
+
+        This is needed when saving optimizers and schedulers in the checkpoint.
```
Comment (stepjam) [inline neuracore/ml/utils/training_storage_handler.py:144]: Why is there checkpoint stuff in this PR. Please split out into another PR.

**Review by stepjam — CHANGES_REQUESTED** [review]
Please split into 3 PRs.

1) checkpoint fixes
2) Adding in support for `configure_optimizer_schedulers`
3) Improving diffusion policy. 

Please make sure to add tests to cover 1 and 2 in their respective PRs

File: examples/ml/example_custom_algorithm.py (hunk 5 above)
Comment (kwangneuraco) [reply examples/ml/example_custom_algorithm.py:635]: > empty list instead of None? Can also add type LRScheduler in the output

Yep I can do empty list. Not all schedulers are LRScheduler and the pi0 scheduler is a custom one with two scheduler mixed, so i am reluctant to add type in this output

File: neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py (hunk 7 above)
Comment (kwangneuraco) [reply neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py:620]: Hi I asked Gemini3 and Claude and both suggest that using `requires_grad=False` is better
image

File: neuracore/ml/utils/training_storage_handler.py (hunk 9 above)
Comment (kwangneuraco) [reply neuracore/ml/utils/training_storage_handler.py:144]: they are used as we save the scheduler parameters. sure another PR is better

File: neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py (hunk 7 above)
Comment (stepjam) [reply neuracore/ml/algorithms/diffusion_policy/diffusion_policy.py:620]: Regardless, requires grad=False would be better placed in init rathe than here
